File: microspice_matlab/+engine/environment.m
% Contains all the components and nodes
classdef environment < matlab.mixin.Copyable
    properties
        num_nodes       int32 = 1;              % The number of nodes
        
        comp_dict       containers.Map;         % Map between component names and their instances
        node_dict       containers.Map;         % Map between node name (incl virtual nodes) to the index
        
        node_vals       containers.Map;         % Value of node variables
        comp_node_idx   containers.Map;         % Map between componenet names and a list of environment indices of each node of the component (for use in stamping)
    end
    
    methods
        % Constructor
        function obj = environment
            obj.comp_dict = containers.Map;
            obj.node_dict = containers.Map(["0"], [1]);
            obj.node_vals = containers.Map(["0"], [0]);
            obj.comp_node_idx = containers.Map;
        end
        
        % Add components to the environment
        function add_component(obj, comp)
            obj.comp_dict(comp.id) = comp;
            
            node_idx_ls = [];
            for node = comp.nodes
                if((~obj.node_dict.isKey(node)))
                    obj.num_nodes = obj.num_nodes + 1;
                    obj.node_dict(node) = obj.num_nodes;
                end
                node_idx_ls(end+1) = obj.node_dict(node);
            end
            obj.comp_node_idx(comp.id) = node_idx_ls;
        end
        
        % Form the matrix to solve DC
        function mat = get_mat_dc(obj)
            stamp_mat = zeros(obj.num_nodes, obj.num_nodes+1);
            
            components = obj.comp_dict.values;
            for i = 1:obj.comp_dict.Count
                comp     = components{i};
                
                index_vec = obj.comp_node_idx(comp.id);
                
                stamp = comp.stamp_dc();
                for j = 1:numel(index_vec)
                    stamp_mat(index_vec(j), [index_vec end]) = stamp_mat(index_vec(j), [index_vec end]) + stamp(j,:);
                end
            end
            
            mat = stamp_mat([2:end], [2:end]);
        end
        
        % Form the matrix to solve transient
        function mat = get_mat_trans(obj, step_size)
            stamp_mat = zeros(obj.num_nodes, obj.num_nodes+1);
            
            components = obj.comp_dict.values;
            for i = 1:obj.comp_dict.Count
                comp     = components{i};
                
                index_vec = obj.comp_node_idx(comp.id);

                stamp = comp.stamp_trans(step_size);
                for j = 1:numel(index_vec)
                    stamp_mat(index_vec(j), [index_vec end]) = stamp_mat(index_vec(j), [index_vec end]) + stamp(j,:);
                end
            end
            
            mat = stamp_mat([2:end], [2:end]);
        end
        
        % Set node values in the node_vals dictionary from the solution vector
        function set_node_vals(obj, soln_vec)
            for n = obj.node_dict.keys
                n = n{1};
                if(n ~= "0")
                    obj.node_vals(n) = soln_vec(obj.node_dict(n)-1);
                end
            end
        end
        
        % Update component states using the node_vals dictionary and time
        function update_comp_states(obj, time)
            components = obj.comp_dict.values;
            
            for i = 1:obj.comp_dict.Count
                comp     = components{i};
                
                comp.update_state(time, obj.node_vals);
            end
        end
    end
    
    methods (Access = protected)        
        % Deep copy (for alter command).
        function cp = copyElement(obj)
            cp = engine.environment;
            for comp = obj.comp_dict.values
                cp.add_component(copy(comp{1}))
            end
        end
    end
end

File: microspice/environment.py
# MIT License

# Copyright (c) 2023 Ashwin Rajesh

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import copy
import numpy as np

class Environment:
    def __init__(self):
        self.num_nodes      = 1             # Total number of nodes

        self.dict_id2comp   = {}            # Map from component ID to component object
        self.dict_node2idx  = {"0": 0}      # Map from node name to node index
        self.dict_node2val  = {"0": 0}      # Map from node name to node value
        self.dict_comp2node = {}            # Map from component ID to nodes connected to the component

    # Add a component to the enviornment
    def add_component(self, comp):
        self.dict_id2comp[comp.id] = comp                   # Add to the component dictionary

        node_idx_ls = []                                    # Maintain a list of nodes connected with the component

        for node in comp.nodes:
            # Add a new node into the enviornment node dictionary
            if node not in self.dict_node2idx:
                self.dict_node2idx[node] = self.num_nodes
                self.num_nodes += 1
            node_idx_ls.append(self.dict_node2idx[node])    # Adding nodes connected to the component
        
        self.dict_comp2node[comp.id] = node_idx_ls          # Add nodes to the component to node dictionary

    # Get the DC matrix to solve
    def get_mat_dc(self):
        stamp_mat = np.zeros((self.num_nodes, self.num_nodes + 1))

        components = self.dict_id2comp.values()
        
        for comp in components:
            idx_vec = self.dict_comp2node[comp.id]
            stamp = comp.stamp_dc()

            # "Stamping"
            for j, idx in enumerate(idx_vec):
                stamp_mat[idx, [idx_vec + [self.num_nodes]]] += stamp[j, :]

        mat = stamp_mat[1:, 1:]
        return mat

    def get_mat_trans(self, step_size):
        stamp_mat = np.zeros((self.num_nodes, self.num_nodes + 1))

        components = self.dict_id2comp.values()
        for comp in components:
            idx_vec = self.dict_comp2node[comp.id]
            stamp   = comp.stamp_trans(step_size)

            # "Stamping" the matrix
            for j, idx in enumerate(idx_vec):
                stamp_mat[idx, [idx_vec + [self.num_nodes]]] += stamp[j, :]

        mat = stamp_mat[1:, 1:]
        return mat

    def set_node_vals(self, soln_vec):
        for node in self.dict_node2idx.keys():
            if node != "0":
                self.dict_node2val[node] = soln_vec[self.dict_node2idx[node] - 1]

    def update_comp_states(self, time):
        components = self.dict_id2comp.values()
        for comp in components:
            comp.update_state(time, self.dict_node2val)

    def copy_element(self):
        cp = Environment()
        for comp in self.dict_id2comp.values():
            cp.add_component(copy.copy(comp))
        return cp
